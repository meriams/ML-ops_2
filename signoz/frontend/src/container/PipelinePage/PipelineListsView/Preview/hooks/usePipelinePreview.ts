import simulatePipelineProcessing, {
	PipelineSimulationResponse,
} from 'api/pipeline/preview';
import { AxiosError } from 'axios';
import { REACT_QUERY_KEY } from 'constants/reactQueryKeys';
import { useQuery } from 'react-query';
import { ILog } from 'types/api/logs/log';
import { PipelineData } from 'types/api/pipeline/def';

export interface PipelinePreviewRequest {
	pipeline: PipelineData;
	inputLogs: ILog[];
}

export interface PipelinePreviewResponse {
	isLoading: boolean;
	outputLogs: ILog[];
	isError: boolean;
	errorMsg: string;
}

const usePipelinePreview = ({
	pipeline,
	inputLogs,
}: PipelinePreviewRequest): PipelinePreviewResponse => {
	// Ensure log timestamps are numbers for pipeline preview API request
	// ILog allows both number and string while the API needs a number
	const simulationInput = inputLogs.map((l) => ({
		...l,
		timestamp: new Date(l.timestamp).getTime(),
	}));

	const response = useQuery<PipelineSimulationResponse, AxiosError>({
		queryFn: async () =>
			simulatePipelineProcessing({
				logs: simulationInput,
				pipelines: [pipeline],
			}),
		queryKey: [REACT_QUERY_KEY.LOGS_PIPELINE_PREVIEW, pipeline, inputLogs],
		retry: false,
	});

	const { isFetching, isError, data, error } = response;

	return {
		isLoading: isFetching,
		outputLogs: data?.logs || [],
		isError,
		errorMsg: error?.response?.data?.error || '',
	};
};

export default usePipelinePreview;
