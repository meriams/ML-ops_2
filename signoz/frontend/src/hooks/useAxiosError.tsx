import axios from 'axios';

import { useNotifications } from './useNotifications';

const useAxiosError = (): UseAxiosError => {
	const { notifications } = useNotifications();

	return (error: unknown): void => {
		if (axios.isAxiosError(error)) {
			notifications.error({
				message: error.message,
			});
		}
	};
};

type UseAxiosError = (error: unknown) => void;

export default useAxiosError;
