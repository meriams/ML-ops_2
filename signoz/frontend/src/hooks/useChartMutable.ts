import { PANEL_TYPES } from 'constants/queryBuilder';
import { PanelTypeAndGraphManagerVisibilityProps } from 'container/GridCardLayout/GridCard/FullView/types';
import { PanelTypeKeys } from 'types/common/queryBuilder';

export const useChartMutable = ({
	panelType,
	panelTypeAndGraphManagerVisibility,
}: UseChartMutableProps): boolean => {
	const panelKeys: PanelTypeKeys[] = [].slice.call(Object.keys(PANEL_TYPES));
	const graphType = panelKeys.find(
		(key: PanelTypeKeys) => PANEL_TYPES[key] === panelType,
	);
	if (!graphType) {
		return false;
	}
	return panelTypeAndGraphManagerVisibility[graphType];
};

interface UseChartMutableProps {
	panelType: string;
	panelTypeAndGraphManagerVisibility: PanelTypeAndGraphManagerVisibilityProps;
}
