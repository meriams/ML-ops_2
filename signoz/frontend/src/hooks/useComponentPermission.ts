import { useCallback, useMemo } from 'react';
import { ROLES } from 'types/roles';
import { componentPermission, ComponentTypes } from 'utils/permission';

const useComponentPermission = (
	component: ComponentTypes[],
	role: ROLES | null,
): boolean[] => {
	const getComponentPermission = useCallback(
		(component: ComponentTypes): boolean =>
			!!componentPermission[component].find((roles) => role === roles),
		[role],
	);

	return useMemo(() => component.map((e) => getComponentPermission(e)), [
		component,
		getComponentPermission,
	]);
};

export default useComponentPermission;
