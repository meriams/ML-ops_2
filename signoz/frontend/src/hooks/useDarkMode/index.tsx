import { theme as antdTheme } from 'antd';
import { ThemeConfig } from 'antd/es/config-provider/context';
import get from 'api/browser/localstorage/get';
import set from 'api/browser/localstorage/set';
import { LOCALSTORAGE } from 'constants/localStorage';
import {
	createContext,
	ReactNode,
	useCallback,
	useContext,
	useMemo,
	useState,
} from 'react';

import { THEME_MODE } from './constant';

export const ThemeContext = createContext({
	theme: THEME_MODE.DARK,
	toggleTheme: () => {},
});

export function ThemeProvider({ children }: ThemeProviderProps): JSX.Element {
	const [theme, setTheme] = useState(get(LOCALSTORAGE.THEME) || THEME_MODE.DARK);

	const toggleTheme = useCallback(() => {
		if (theme === THEME_MODE.LIGHT) {
			setTheme(THEME_MODE.DARK);
			set(LOCALSTORAGE.THEME, THEME_MODE.DARK);
		} else {
			setTheme(THEME_MODE.LIGHT);
			set(LOCALSTORAGE.THEME, THEME_MODE.LIGHT);
		}
	}, [theme]);

	const value = useMemo(
		() => ({
			theme,
			toggleTheme,
		}),
		[theme, toggleTheme],
	);

	return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
}

interface ThemeProviderProps {
	children: ReactNode;
}

interface ThemeMode {
	theme: string;
	toggleTheme: () => void;
}

export const useThemeMode = (): ThemeMode => {
	const { theme, toggleTheme } = useContext(ThemeContext);

	return { theme, toggleTheme };
};

export const useIsDarkMode = (): boolean => {
	const { theme } = useContext(ThemeContext);

	return theme === THEME_MODE.DARK;
};

export const useThemeConfig = (): ThemeConfig => {
	const isDarkMode = useIsDarkMode();

	return {
		algorithm: isDarkMode ? antdTheme.darkAlgorithm : antdTheme.defaultAlgorithm,
		token: {
			borderRadius: 2,
			borderRadiusLG: 2,
			borderRadiusSM: 2,
			borderRadiusXS: 2,
			fontFamily: 'Inter',
			fontSize: 13,
		},
	};
};

export default useThemeMode;
