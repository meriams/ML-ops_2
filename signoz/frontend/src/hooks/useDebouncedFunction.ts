import debounce from 'lodash-es/debounce';
import { useMemo, useRef } from 'react';

export interface DebouncedFunc<T extends (...args: unknown[]) => unknown> {
	(...args: Parameters<T>): ReturnType<T> | undefined;

	cancel(): void;

	flush(): ReturnType<T> | undefined;
}

export type DebounceOptions = {
	leading?: boolean;
	maxWait?: number;
	trailing?: boolean;
};

const defaultOptions: DebounceOptions = {
	leading: false,
	trailing: true,
};

const useDebouncedFn = <T extends (...args: Array<unknown>) => unknown>(
	fn: T,
	wait = 100,
	options: DebounceOptions = defaultOptions,
): DebouncedFunc<T> => {
	const fnRef = useRef(fn);
	fnRef.current = fn;

	return useMemo(
		() => debounce(((...args) => fnRef.current(...args)) as T, wait, options),
		[options, wait],
	);
};

export default useDebouncedFn;
