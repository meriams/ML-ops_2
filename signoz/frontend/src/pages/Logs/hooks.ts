// utils
import get from 'api/browser/localstorage/get';
import { LOCALSTORAGE } from 'constants/localStorage';
// interfaces
import { LogViewMode } from 'container/LogsTable';
import { useCallback, useLayoutEffect, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { setLinesPerRow } from 'store/actions/logs/setLInesPerRow';
// actions
import { setViewMode } from 'store/actions/logs/setViewMode';
import { AppState } from 'store/reducers';

import { viewModeOptionList } from './config';
import { SelectedLogViewData } from './types';
import { isLogViewMode } from './utils';

export const useSelectedLogView = (): SelectedLogViewData => {
	const dispatch = useDispatch();
	const viewMode = useSelector<AppState, LogViewMode>(
		(state) => state.logs.viewMode,
	);
	const linesPerRow = useSelector<AppState, number>(
		(state) => state.logs.linesPerRow,
	);

	const viewModeOption = useMemo(
		() =>
			viewModeOptionList.find(
				(viewModeOption) => viewModeOption.value === viewMode,
			) ?? viewModeOptionList[0],
		[viewMode],
	);

	const handleViewModeChange = useCallback(
		(selectedViewMode: LogViewMode) => {
			dispatch(setViewMode(selectedViewMode));
		},
		[dispatch],
	);

	const handleViewModeOptionChange = useCallback(
		({ key }: { key: string }) => {
			if (isLogViewMode(key)) handleViewModeChange(key);
		},
		[handleViewModeChange],
	);

	const handleLinesPerRowChange = useCallback(
		(selectedLinesPerRow: unknown) => {
			if (typeof selectedLinesPerRow === 'number') {
				dispatch(setLinesPerRow(selectedLinesPerRow));
			}
		},
		[dispatch],
	);

	useLayoutEffect(() => {
		const storedViewMode = get(LOCALSTORAGE.LOGS_VIEW_MODE);
		if (storedViewMode) {
			handleViewModeChange(storedViewMode as LogViewMode);
		}

		const storedLinesPerRow = get(LOCALSTORAGE.LOGS_LINES_PER_ROW);
		if (storedLinesPerRow) {
			handleLinesPerRowChange(+storedLinesPerRow);
		}
	}, [handleViewModeChange, handleLinesPerRowChange]);

	return {
		viewModeOptionList,
		viewModeOption,
		viewMode,
		handleViewModeChange,
		handleViewModeOptionChange,
		linesPerRow,
		handleLinesPerRowChange,
	};
};
