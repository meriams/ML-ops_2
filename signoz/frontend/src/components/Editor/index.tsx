import MEditor, { EditorProps } from '@monaco-editor/react';
import { useIsDarkMode } from 'hooks/useDarkMode';
import { useMemo } from 'react';

function Editor({
	value,
	language,
	onChange,
	readOnly,
	height,
	options,
}: MEditorProps): JSX.Element {
	const isDarkMode = useIsDarkMode();

	const onChangeHandler = (newValue?: string): void => {
		if (readOnly) return;

		if (typeof newValue === 'string' && onChange) onChange(newValue);
	};

	const editorOptions = useMemo(
		() => ({ fontSize: 16, automaticLayout: true, readOnly, ...options }),
		[options, readOnly],
	);

	return (
		<MEditor
			theme={isDarkMode ? 'vs-dark' : 'vs-light'}
			language={language}
			value={value}
			options={editorOptions}
			height={height}
			onChange={onChangeHandler}
			data-testid="monaco-editor"
		/>
	);
}

interface MEditorProps {
	value: string;
	language?: string;
	onChange?: (value: string) => void;
	readOnly?: boolean;
	height?: string;
	options?: EditorProps['options'];
}

Editor.defaultProps = {
	language: 'yaml',
	readOnly: false,
	height: '40vh',
	options: {},
	onChange: (): void => {},
};

export default Editor;
