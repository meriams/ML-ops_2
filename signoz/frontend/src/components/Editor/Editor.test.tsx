import { render, screen } from '@testing-library/react';
import { useIsDarkMode } from 'hooks/useDarkMode';

import Editor from './index';

jest.mock('hooks/useDarkMode', () => ({
	useIsDarkMode: jest.fn(),
}));

describe('Editor', () => {
	it('renders correctly with default props', () => {
		const { container } = render(<Editor value="" />);
		expect(container).toMatchSnapshot();
	});

	it('renders correctly with custom props', () => {
		const customProps = {
			value: 'test',
			language: 'javascript',
			readOnly: true,
			height: '50vh',
			options: { minimap: { enabled: false } },
		};
		const { container } = render(
			<Editor
				value={customProps.value}
				height={customProps.height}
				language={customProps.language}
				options={customProps.options}
				readOnly={customProps.readOnly}
			/>,
		);
		expect(container).toMatchSnapshot();
	});

	it('renders with dark mode theme', () => {
		(useIsDarkMode as jest.Mock).mockImplementation(() => true);

		const { container } = render(<Editor value="dark mode text" />);

		expect(container).toMatchSnapshot();
	});

	it('renders with light mode theme', () => {
		(useIsDarkMode as jest.Mock).mockImplementation(() => false);

		const { container } = render(<Editor value="light mode text" />);

		expect(container).toMatchSnapshot();
	});

	it('displays "Loading..." message initially', () => {
		const { rerender } = render(<Editor value="initial text" />);

		expect(screen.getByText('Loading...')).toBeInTheDocument();

		rerender(<Editor value="initial text" />);
	});
});
