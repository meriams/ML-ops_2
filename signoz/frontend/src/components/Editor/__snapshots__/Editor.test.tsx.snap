// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor renders correctly with custom props 1`] = `
<div>
  <section
    style="display: flex; position: relative; text-align: initial; width: 100%; height: 50vh;"
  >
    <div
      style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
    >
      Loading...
    </div>
    <div
      style="width: 100%; display: none;"
    />
  </section>
</div>
`;

exports[`Editor renders correctly with default props 1`] = `
<div>
  <section
    style="display: flex; position: relative; text-align: initial; width: 100%; height: 40vh;"
  >
    <div
      style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
    >
      Loading...
    </div>
    <div
      style="width: 100%; display: none;"
    />
  </section>
</div>
`;

exports[`Editor renders with dark mode theme 1`] = `
<div>
  <section
    style="display: flex; position: relative; text-align: initial; width: 100%; height: 40vh;"
  >
    <div
      style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
    >
      Loading...
    </div>
    <div
      style="width: 100%; display: none;"
    />
  </section>
</div>
`;

exports[`Editor renders with light mode theme 1`] = `
<div>
  <section
    style="display: flex; position: relative; text-align: initial; width: 100%; height: 40vh;"
  >
    <div
      style="display: flex; height: 100%; width: 100%; justify-content: center; align-items: center;"
    >
      Loading...
    </div>
    <div
      style="width: 100%; display: none;"
    />
  </section>
</div>
`;
